python-frontend/src/main/java/org/sonar/python/semantic/ProjectLevelSymbolTable.java

```diff
@@ -56,75 +59,107 @@ public static ProjectLevelSymbolTable from(Map<String, Set<Symbol>> globalSymbol
+        globalDescriptors.addAll(DescriptorUtils.descriptors(symbol));
       }
     }
-    globalSymbolsByModuleName.put(fullyQualifiedModuleName, globalSymbols);
-    if (globalSymbolsByFQN != null) {
+    globalDescriptorsByModuleName.put(fullyQualifiedModuleName, globalDescriptors);
+    if (globalDescriptorsByFQN != null) {
       // TODO: build globalSymbolsByFQN incrementally
-      addModuleToGlobalSymbolsByFQN(globalSymbols);
+      addModuleToGlobalSymbolsByFQN(globalDescriptors);
     }
     DjangoViewsVisitor djangoViewsVisitor = new DjangoViewsVisitor();
     fileInput.accept(djangoViewsVisitor);
   }
 
-  private void addModuleToGlobalSymbolsByFQN(Set<Symbol> symbols) {
-    Map<String, Symbol> moduleSymbolsByFQN = symbols.stream()
-      .filter(symbol -> symbol.fullyQualifiedName() != null)
-      .collect(Collectors.toMap(Symbol::fullyQualifiedName, Function.identity(), AmbiguousSymbolImpl::create));
-    globalSymbolsByFQN.putAll(moduleSymbolsByFQN);
+  private void addModuleToGlobalSymbolsByFQN(Set<Descriptor> descriptors) {
+    Map<String, Descriptor> moduleDescriptorsByFQN = descriptors.stream()
+      .filter(d -> d.fullyQualifiedName() != null)
+      .collect(Collectors.toMap(Descriptor::fullyQualifiedName, Function.identity(), AmbiguousDescriptor::create));
+    globalDescriptorsByFQN.putAll(moduleDescriptorsByFQN);
+
   }
 
-  private Map<String, Symbol> globalSymbolsByFQN() {
-    if (globalSymbolsByFQN == null) {
-      globalSymbolsByFQN = globalSymbolsByModuleName.values()
+  private Map<String, Descriptor> globalDescriptorsByFQN() {
+    if (globalDescriptorsByFQN == null) {
+      globalDescriptorsByFQN = globalDescriptorsByModuleName.values()
         .stream()
         .flatMap(Collection::stream)
-        .filter(symbol -> symbol.fullyQualifiedName() != null)
-        .collect(Collectors.toMap(Symbol::fullyQualifiedName, Function.identity(), AmbiguousSymbolImpl::create));
+        .filter(descriptor -> descriptor.fullyQualifiedName() != null)
+        .collect(Collectors.toMap(Descriptor::fullyQualifiedName, Function.identity(), AmbiguousDescriptor::create));
     }
-    return globalSymbolsByFQN;
+    return globalDescriptorsByFQN;
   }
 
   @CheckForNull
   public Symbol getSymbol(@Nullable String fullyQualifiedName) {
-    return globalSymbolsByFQN().get(fullyQualifiedName);
+    return getSymbol(fullyQualifiedName, null);
+  }
+
+  public Symbol getSymbol(@Nullable String fullyQualifiedName, @Nullable String localSymbolName) {
+    if (fullyQualifiedName == null) return null;
+    if (queriedSymbolNames.contains(fullyQualifiedName)) {
+      // cyclic dependencies
+      queriedSymbolNames = new HashSet<>();
+      String[] fqnSplitByDot = fullyQualifiedName.split("\\.");
+      localSymbolName =  localSymbolName != null ? localSymbolName : fqnSplitByDot[fqnSplitByDot.length - 1];
+      return new SymbolImpl(localSymbolName, fullyQualifiedName);
+    }
+    queriedSymbolNames.add(fullyQualifiedName);
```

Comment:
I agree I don't think it's as elegant as it could be.
The idea is the following:
- If we hit the reset of line `138`, we have an infinite recursion and we just return a placeholder symbol.
- If we hit the reset of line `146`, that means we don't have a descriptor for the FQN and we'll just return null. There's no longer a risk of recursion so we can reset safely.
- The call DescriptorUtils#symbolFromDescriptor might indeed recurse as it can involve reconstructing type hierarchies by fetching symbols from FQNs. Once we hit line `156`, we've either successfully reconstructed the symbol without recursion, or short circuited it by returning a placeholder symbol. In both cases, we want to make sure `queriedSymbolNames` is cleared for future, unrelated searches.

Maybe it might be cleared to add the fully qualified name just before the `DescriptorUtils#symbolFromDescriptor` call as it's the only one at risk of creating infinite recursion. If you have other suggestions to make this a bit cleaner, please let me know.